### PR TITLE
Feature/i18n chuncks interpolator

### PR DESCRIPTION
**Pull request by tonyghiani**
Implement a `I18n.prototype.t` for improved boudaries interpolation.

## Description
In some exceptional cases, we may want to parse our translations with a stronger interpolation algorithm. For instance, we may want to collect some text between specific boundaries, and we want to return a specific result for that specific chunk of text.

A typical use case is when a literal contains special styles or transformations in a React component, like an anchor tag in the middle of the text or anything else that would require enhancing a portion of text.

`I18n.prototype.interpolate` has all the features of `I18n.prototype.t` and provides the additional flexibility of collecting the children's content between boundaries, so that we can return the desired value.

It **DOES NOT** return a string anymore, instead, it'll give back an array of chunks.

The parser supports:

- Multiple sequential groups.
- Nested interpolation

```javascript
import I18n from '@s-ui/i18n'
import Polyglot from '@s-ui/i18n/lib/adapters/polyglot'
const i18n = new I18n({adapter: new Polyglot()})
i18n.languages = {
  'en-GB': {
    SIMPLE_BOUNDARY: 'Your name is %[upper]Spike[upper]%, the best superhero!',
    SIMPLE_BOUNDARY_PLACEHOLDER:
      '%{name}, your name is %[upper]%{name}[upper]%!',
    MULTIPLE_BOUNDARY:
      '%[upper]%{name}[upper]%, your name is %[upper]%{name}[upper]%!',
    NESTED_PLACEHOLDERS:
      'Hello, this is an %[bold]important %[link]resource[link]%[bold]% for this page!'
  }
}
i18n.culture = 'es-ES'
i18n.interpolate('SIMPLE_BOUNDARY', {
  name: 'Spike',
  upper: ({children}) => children.toUpperCase()
}) //=> ['Spike, your name is ', 'SPIKE', '!']
i18n.interpolate('SIMPLE_BOUNDARY_PLACEHOLDER', {
  name: 'Spike',
  upper: ({children}) => children.toUpperCase()
}) //=> ['Spike, your name is ', 'SPIKE', '!']
i18n.interpolate('MULTIPLE_BOUNDARY', {
  name: 'Spike',
  upper: ({children}) => children.toUpperCase()
}) //=> ['SPIKE', ', your name is ', 'SPIKE', '!']
i18n.interpolate('NESTED_PLACEHOLDERS', {
  bold: ({children}) => <strong>{children}</strong>,
  link: ({children}) => <a>{children}</a>,
}) //=> ['Hello, this is an ', <strong>['important ', <a>resource</a>]</strong>,' for this page!']
```

## Example
The `I18n.prototype.interpolate` method is particularly helpful to generate React.Fragments that can be rendered inside a component.

```jsx
import I18n from '@s-ui/i18n'
import Polyglot from '@s-ui/i18n/lib/adapters/polyglot'
const i18n = new I18n({adapter: new Polyglot()})
i18n.languages = {
  'en-GB': {
    DISCLAIMER: 'Complete cvs have %[gray]twice the change to be picked.[gray]%. Start now completing it!'
  }
}
function Component () {
  const label = i18n.interpolate(`DISCLAIMER`, {
    gray: props => <Text as="span" color={COLORS['gray-D1']} {...props} />
  })
  
  return <Text>{label}</Text>
}
```

### CODE REVIEW

File: packages/sui-i18n/README.md [hunk 1]
```diff
@@ -59,43 +60,134 @@ i18n.t('PLURAL_WITH_VARIABLE', {houses: 2}) //=> 2 casas
...
+    NESTED_PLACEHOLDERS:
+      'Hello, this is an %[bold]important %[link]resource[link]%[bold]% for this page!'
+  }
+}
+
+i18n.culture = 'es-ES'
+i18n.interpolate('SIMPLE_BOUNDARY', {
+  name: 'Spike',
+  upper: ({children}) => children.toUpperCase()
+}) //=> ['Spike, your name is ', 'SPIKE', '!']
```
Comment (rmoralp) [inline packages/sui-i18n/README.md:119]:
```suggestion
}) //=> ['Spike, your name is ', 'SPIKE', ', the best superhero!']
```

File: packages/sui-i18n/README.md [hunk 2]
```diff
@@ -59,43 +60,134 @@ i18n.t('PLURAL_WITH_VARIABLE', {houses: 2}) //=> 2 casas
...
+i18n.interpolate('SIMPLE_BOUNDARY_PLACEHOLDER', {
+  name: 'Spike',
+  upper: ({children}) => children.toUpperCase()
+}) //=> ['Spike, your name is ', 'SPIKE', '!']
+i18n.interpolate('MULTIPLE_BOUNDARY', {
+  name: 'Spike',
+  upper: ({children}) => children.toUpperCase()
+}) //=> ['SPIKE', ', your name is ', 'SPIKE', '!']
```
Comment (rmoralp) [inline packages/sui-i18n/README.md:127]: Could you fix this comment with correct value? Also, line 123 is not correct

**Review by rmoralp — APPROVED** [review]

**Comment by rmoralp** [comment]
I have found to be missing a best practices section in readme. I mean... What kind of boundaries content are a good and a bad practice? Is it OK to interpolate an entire react component like a card in a literal?

**Comment by tonyghiani** [comment]
> I have found to be missing a best practices section in readme. I mean... What kind of boundaries content are a good and a bad practice? Is it OK to interpolate an entire react component like a card in a literal?

I agree, I'll add a section with DO/DON'T and I'll ping you again to check it out :)

File: packages/sui-i18n/README.md (hunk 2 above)
Comment (tonyghiani) [reply packages/sui-i18n/README.md:127]: Mmm I don't see the error, `SIMPLE_BOUNDARY_PLACEHOLDER` and `MULTIPLE_BOUNDARY` are different from the `SIMPLE_BOUNDARY` copy

File: packages/sui-i18n/README.md [hunk 3]
```diff
@@ -267,32 +361,215 @@ i18n.url('/HOME') //=> 'www.example.com/home'
...
+const i18n = new I18n({adapter: new Polyglot()})
+i18n.languages = {
+  'en-GB': {
+    TEXT_KEY: 'Complete your data!'
+  }
+}
+
+// ✅ Do
+function Component() {
+  const label = i18n.t(`TEXT_KEY`)
```
Comment (andresz1) [inline packages/sui-i18n/README.md:404]: Instead of using ` here maybe would be better to use '. e.g. 'TEXT_KEY'

Comment (tonyghiani) [reply packages/sui-i18n/README.md:404]: It's a leftover from other examples, good catch, let me fix it :)

**Review by andresz1 — APPROVED** [review]

**Pull request merged**